src/coreclr/src/vm/eventtrace.cpp

```diff
@@ -3498,75 +3498,87 @@ BOOL ETW::TypeSystemLog::AddTypeToGlobalCacheIfNotExists(TypeHandle th, BOOL * p
+    if (pLoggedTypesFromModule == NULL)
+    {
+        pLoggedTypesFromModule = new (nothrow) LoggedTypesFromModule(pLoaderModule);
         if (pLoggedTypesFromModule == NULL)
         {
-            pLoggedTypesFromModule = new (nothrow) LoggedTypesFromModule(pLoaderModule);
-            if (pLoggedTypesFromModule == NULL)
-            {
-                // out of memory.  Bail on ETW stuff
-                *pfCreatedNew = FALSE;
-                return fSucceeded;
-            }
-
-            fSucceeded = FALSE;
-            EX_TRY
-            {
-                s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule);
-                fSucceeded = TRUE;
-            }
-            EX_CATCH
-            {
-                fSucceeded = FALSE;
-            }
-            EX_END_CATCH(RethrowCorruptingExceptions);
-            if (!fSucceeded)
-            {
-                *pfCreatedNew = FALSE;
-                return fSucceeded;
-            }
+            // out of memory.  Bail on ETW stuff
+            *pfCreatedNew = FALSE;
+            return fSucceeded;
         }
+    }
 
-        // Step 2: From hash of types, see if our TypeHandle is there already
-        TypeLoggingInfo typeLoggingInfoPreexisting = pLoggedTypesFromModule->loggedTypesFromModuleHash.Lookup(th);
+    fSucceeded = FALSE;
+    EX_TRY
+    {
+        CrstHolder _crst(GetHashCrst());
+        s_pAllLoggedTypes->allLoggedTypesHash.Add(pLoggedTypesFromModule);
+        fSucceeded = TRUE;
+    }
+    EX_CATCH
+    {
+        fSucceeded = FALSE;
+    }
+    EX_END_CATCH(RethrowCorruptingExceptions);
+    if (!fSucceeded)
+    {
+        *pfCreatedNew = FALSE;
+        return fSucceeded;
+    }
+
+    // Step 2: From hash of types, see if our TypeHandle is there already
+    TypeLoggingInfo typeLoggingInfoPreexisting;
+
+    {
+        CrstHolder _crst(GetHashCrst());
+        typeLoggingInfoPreexisting = pLoggedTypesFromModule->loggedTypesFromModuleHash.Lookup(th);
         if (!typeLoggingInfoPreexisting.th.IsNull())
         {
             // Type is already hashed, so it's already logged, so we don't need to
             // log it again.
             *pfCreatedNew = FALSE;
             return fSucceeded;
         }
+    }
 
-        // We haven't logged this type, so we need to continue with this function to
-        // log it below. Add it to the hash table first so any recursive calls will
-        // see that this type is already being taken care of
+    // We haven't logged this type, so we need to continue with this function to
+    // log it below. Add it to the hash table first so any recursive calls will
+    // see that this type is already being taken care of
+    fSucceeded = FALSE;
+    TypeLoggingInfo typeLoggingInfoNew(th);
+    EX_TRY
+    {
+        CrstHolder _crst(GetHashCrst());
+        pLoggedTypesFromModule->loggedTypesFromModuleHash.Add(typeLoggingInfoNew);
```

Comment:
Same issue as above, another Lookup() is necessary to ensure that a 2nd thread has not already added this item while the lock wasn't held.